### ISSUE TITLE
Implement some sort of `.and` higher order function to accompany `.map`

### ISSUE THREAD

**Issue by zesterer** [yoanlcq/vek#14]
There are a lot of use-cases for performing an operation on 2 vectors on an element-by-element basis.

`vek` already supports this through the built-in arithmetic operator overloads.

This should be extended further in a similar manner to `std`'s `Option::and_then` to allow things like this:

```rust
let va = Vec3::new(3, -24, 32);
let vb = Vec3::new(4, 9, 13);

let vc = v0.and(|a, b| a.div_euc(b));
```

**Comment by yoanlcq** [comment]
Yes, this would be convenient. This reminds me of [the `map2()` method of the `Pixel` trait of the `image` crate](https://docs.rs/image/0.20.0/image/trait.Pixel.html#tymethod.map2).  
While we're at it, we could also add `apply()` and `apply2()`, which are the same as map but operate in-place.

About the name, I would prefer `map2` instead of `and`, because `and` reminds too much of the logical operator (and could be confused with `reduce_and()`).

For consistency, this should also be implemented on matrices. Not sure about other structures.

I suspect there are typos in your sample code. Just to be sure, did you rather mean something like this ?

```rust
let va = Vec3::new(...);
let vb = Vec3::new(...);
let vc = va.and(vb, |a, b| a.div_euc(b)); // <-- corrected?
```

If you're okay with all of this, I'll start to work on it; looks easy enough. :)

**Comment by zesterer** [comment]
Yes, something exactly like that. Apologies for derping and missing the `vb` argument.

Looking forward to this! Do you know when it might be done by? We have a few use-cases in Veloren.

As an off-topic side note, I thought you might like to see what your library is being used for (it's now the only vector/matrix maths library in the codebase):
![image](https://user-images.githubusercontent.com/3268430/47381787-b3b4bc80-d6f8-11e8-8ec7-176b6dff542c.png)

**Comment by yoanlcq** [comment]
Wooo! Looking great! :) Thanks for sharing, it's very inspiring and I'm glad my library could help.

I was planning to start implementing `map2()` and friends this evening. Should be pretty quick, actually the longest part should be, triple-checking that everything is OK, compiling and pushing a new version.

**Comment by yoanlcq** [comment]
Well, that should do it; implemented in 0.9.5. :)
